rust libnm_dbus: Preserve unsupported options in NmConnection

cathay4t opened the pull request:
When serializing/deserializing, we should also include unsupported
options in NmConnection and its subordinates(NmSettingXxx).

We use `_other` property to hold unknown data in NmConnection and
NmSettingXxx.

almusil commented on rust/src/libnm_dbus/connection/bridge.rs:
```diff
@@ -16,51 +16,76 @@
 #[derive(Debug, Clone, PartialEq, Default)]
 pub struct NmSettingBridge {
     pub stp: Option<bool>,
+    _other: HashMap<String, zvariant::OwnedValue>,
```
nit: underscore is really for something that is not used at all but here it is at least populated.

cathay4t replied:
The reason I use `_other` instead of `other` because the `other` might be used by NetworkManager as property name someday, I don't want to conflict with that and be inconsistent on all `NmSettingXxx`.

And of course, it is also my old C habit.

almusil commented on rust/src/libnm_dbus/connection/bridge.rs:
```diff
@@ -16,51 +16,76 @@
+                    setting._other.insert(key.to_string(), value.clone());
+                }
+            }
+        }
```
It could be shortened to something like:
```
let mut setting = Self::new();
let mut setting_cloned = setting_value.clone();
if let Some(stp_val) = setting_cloned.remove("stp") {
    setting.stp = Some(own_value_to_bool(&stp_val)?);
}
setting._other = setting_cloned;
```

I am not sure if that's desirable.

Maybe it would be worth to have some similar  helper to removed `value_hash_get_bool`. 
E.g. `value_hash_remove_bool` that would simplify this whole thing into:

```
let mut setting = Self::new();
let mut setting_cloned = setting_value.clone();
setting.stp = value_hash_remove_bool(&setting_cloned, "stp")?;
setting._other = setting_cloned;
```

Which is more readable IMO.

cathay4t replied:
In rust, I intend to refrain myself from using `clone()` whenever possible. Cloning the whole setting seems against this.

almusil replied:
That's a valid point but the whole map is cloned anyway, in this case not cloning everything spares probably only the stp string from being cloned. Also another point might be that by cloning everything there is a single allocation happening instead of allocate steps during insert (That could be avoided by pre-allocating the _other).

cathay4t replied:
If we do clone there, we will do cloning again when `own_value_to_string()` and others.

almusil replied:
That could be probably converted to something like:
```
pub(crate) fn own_value_to_string(
    value: zvariant::OwnedValue,
) -> Result<String, NmError> {
    check_value_is_string(&value)?;
    String::try_from(value).map_err(|e| {
        NmError::new(
            ErrorKind::Bug,
            format!("Failed to convert {:?} to string: {}", &value, e),
        )
    })
}
```

Yeah I know that cloning can be endless fight in rust

cathay4t replied:
New patch uploaded using this big clone() and extract memory from OwnValue out.

almusil commented on rust/src/libnm_dbus/connection/bridge.rs:
```diff
@@ -16,51 +16,76 @@
         }
+        for (key, value) in &self._other {
+            ret.insert(key.as_str(), zvariant::Value::from(value.clone()));
+        }
```
Could be:
```
ret.extend(self._other.iter().map(|(key, value)| {
    (key.as_str(), zvariant::Value::from(value.clone()))
}));
```

But I guess here it is more about preference.

cathay4t replied:
The code is better. Will change to that.

almusil commented on rust/src/libnm_dbus/keyfile.rs:
```diff
@@ -0,0 +1,186 @@
+            for item in a.get() {
+                ret += &zvariant_value_to_keyfile(item, section_name)?;
+                ret += ";";
+                //ret += &format!("{}={}\n", prop_name, d);
```
Leftover?

cathay4t replied:
Removed. Thanks.

almusil reviewed (commented):
Hi,

hopefully you don't mind me doing review. 

Those patterns that I have pointed out are common through this whole change.
IMO it would make it more readable but leaving it up to you.

almusil commented on rust/src/libnm_dbus/connection/conn.rs:
```diff
@@ -65,35 +63,48 @@ impl TryFrom<NmConnectionDbusOwnedValue> for NmConnection {
+                        .insert(key.to_string(), setting_value.clone());
+                }
+            }
         }
```
Looking at it, we could avoid most of the cloning by passing it from here as a owned value. If you don't mind I can create a commit on top of this one to show you what I mean.

almusil commented on rust/src/libnm_dbus/connection/conn.rs:
```diff
@@ -56,44 +53,58 @@ pub struct NmConnection {
+                        ._other
+                        .insert(key.to_string(), setting_value.clone());
+                }
+            }
```
Is this intentionally left as iterator? This could also be the remove.map.transpose combination

cathay4t replied:
I forgot that one. Fixed.

cathay4t committed:
rust libnm_dbus: Preserve unsupported options in NmConnection

When serializing/deserializing, we should also include unsupported
options in NmConnection and its subordinates(NmSettingXxx).

We use `_other` property to hold unknown data in NmConnection and
NmSettingXxx.

Also changed to code to save the memory duplication by converting
OwnedValue to String/u32/etc directly without clone.

Signed-off-by: Gris Ge <fge@redhat.com>

almusil reviewed (approved):
Looks good
